model/riscv_insts_zbkb.sail

```diff
@@ -0,0 +1,104 @@
+/* ****************************************************************** */
+union clause ast = ZBKB_RTYPE : (regidx, regidx, regidx, brop_zbkb)
+
+mapping clause encdec = ZBKB_RTYPE(rs2, rs1, rd, RISCV_PACK) if haveZbkb()
+  <-> 0b0000100 @ rs2 @ rs1 @ 0b100 @ rd @ 0b0110011 if haveZbkb()
+
+mapping clause encdec = ZBKB_RTYPE(rs2, rs1, rd, RISCV_PACKH) if haveZbkb()
+  <-> 0b0000100 @ rs2 @ rs1 @ 0b111 @ rd @ 0b0110011 if haveZbkb()
+
+mapping zbkb_rtype_mnemonic : brop_zbkb <-> string = {
+  RISCV_PACK  <-> "pack",
+  RISCV_PACKH <-> "packh"
+}
+
+mapping clause assembly = ZBKB_RTYPE(rs2, rs1, rd, op)
+  <-> zbkb_rtype_mnemonic(op) ^ spc() ^ reg_name(rd) ^ sep() ^ reg_name(rs1) ^ sep() ^ reg_name(rs2)
+
+function clause execute (ZBKB_RTYPE(rs2, rs1, rd, op)) = {
+  let rs1_val = X(rs1);
+  let rs2_val = X(rs2);
+  let result : xlenbits = match op {
+    RISCV_PACK  => if sizeof(xlen) == 32
+                   then rs2_val[15..0] @ rs1_val[15..0]
+                   else rs2_val[31..0] @ rs1_val[31..0],
+    RISCV_PACKH => EXTZ(rs2_val[7..0] @ rs1_val[7..0])
+  };
+  X(rd) = result;
+  RETIRE_SUCCESS
+}
+
+/* ****************************************************************** */
+union clause ast = ZBKB_PACKW : (regidx, regidx, regidx)
+
+mapping clause encdec = ZBKB_PACKW(rs2, rs1, rd) if haveZbkb() & sizeof(xlen) == 64
+  <-> 0b0000100 @ rs2 @ rs1 @ 0b100 @ rd @ 0b0111011 if haveZbkb() & sizeof(xlen) == 64
+
+mapping clause assembly = ZBKB_PACKW(rs2, rs1, rd)
+  <-> "packw" ^ spc() ^ reg_name(rd) ^ sep() ^ reg_name(rs1) ^ sep() ^ reg_name(rs2)
+
+function clause execute (ZBKB_PACKW(rs2, rs1, rd)) = {
+  let rs1_val = X(rs1);
+  let rs2_val = X(rs2);
+  let result : bits(32) = rs2_val[15..0] @ rs1_val[15..0];
+  X(rd) = EXTZ(result);
+  RETIRE_SUCCESS
+}
+
+/* ****************************************************************** */
+union clause ast = RISCV_ZIP : (regidx, regidx)
+
+mapping clause encdec = RISCV_ZIP(rs1, rd) if haveZbkb() & sizeof(xlen) == 32
+  <-> 0b000010001111 @ rs1 @ 0b001 @ rd @ 0b0010011 if haveZbkb() & sizeof(xlen) == 32
+
+mapping clause assembly = RISCV_ZIP(rs1, rd)
+  <-> "zip" ^ spc() ^ reg_name(rd) ^ sep() ^ reg_name(rs1)
+
+function clause execute (RISCV_ZIP(rs1, rd)) = {
+  let rs1_val = X(rs1);
+  result : xlenbits = zeros();
```

Comment:
Either consistently hard-code that it's RV32 or write fully general code that happens to only run on RV32 (i.e., either change xlenbits to bits(32) here or change the 15 and 16 below to be `sizeof(xlen)/2` and that `- 1`). If doing the former you will want an `assert(sizeof(xlen) == 32);` at the start of this function to inform the type checker it's dead code on RV64.

For what it's worth, the pseudo-Sail in the the crypto spec takes the "fully general code that happens to only run on RV32" approach, but I think both are ok, doesn't matter unless there's a reason to believe a future extension might want (un)zip on RV64.

Reply:
sure.

Reply:
Since division is not available I'll hard-code xlenbits to bits(32)

Reply:
There's also `sizeof(xlen_bytes) * 4` as an option